### DISCUSSION TITLE
Postgres: warning for  not unique __text values in key/value variable feature.

### DISCUSSION THREAD

**Discussion by gjed** [grafana/grafana#50508] [Data Sources]
## What would you like to be added
Some kind of warning/support in the edit variable UI (maybe above or below the values preview), when the selected __text values in a key/value variable are not unique.
Example with crappy css:

![warning](https://user-images.githubusercontent.com/12700875/172838011-45d40a8c-b3c9-4e8f-b555-a2be5a379489.png)

## Why is this needed
In my opinion, while I fully agree that the correctness of the query is a burden that falls on the query writer, it would be really helpful to have some kind of support for this situation.
The documentation is actually clear about this but it is not intuitive from a sql point of view (id/primary key is unique while label might not be)

## My case

I'll start saying that the [documentation](https://grafana.com/docs/grafana/latest/datasources/postgres/#query-variable) clearly says and i quote:
> The query should return two columns that are named `__text` and `__value`. The `__text` column value should be unique (if it is not unique then the first value is used). 

I found this out the hard way: while building a new dashboard (for kpi porpuses) i had a few panels for which the displayed data was not coherent. The queries involved different tables and where not as straightforward as a simple time series.
In particular, there was a hierarchical structure, which can be summarized as follow:
- there are object of type A
- there are object of type B that are in a 1-1 relationship with an object of type A
- there are object of type C that are in a 1-1 relationship with an object of type C
- the data is contained in the lowest element ( of type C in this case)
- the data must be queryable by A and B (for example: see the data of type C objects that are related to A1)

What was happening was that, since some B and C objects had different id but same label, as described in the docs, only the first occurence of `__text` was shown. The two node graph panels show an example of the difference of results between the same query with variables with unique `__text` and variables with duplicated `__text` value:

![Screenshot from 2022-06-09 13-39-21](https://user-images.githubusercontent.com/12700875/172838234-d33cbeaa-bee3-4639-85c7-2fce21ab7e3a.png)

All this to say that, even though the documentation is clear, it is not easy to debug this situation when writing the queries.

It's a long description for a small and very specific situation, but I think it would be helpful.
Here are the [sql](https://gist.github.com/gjed/9178e84b5f3088f50b5e4cb95e239b22.js) and the [dashboard json](https://gist.github.com/gjed/8e1dfecf3fd63d0df4d794902491f661) to replicate the situation. The two node graph panels show the difference of results between the same query with variables with unique `__text` and variables with duplicated `__text` value

Thansk!

gjed

**Comment by gabor** [comment]
hi, i converted this to a github issue at https://github.com/grafana/grafana/issues/78403. closing this discussion.
